server/spec/services/mongo_pubsub_spec.rb

```diff
@@ -96,15 +96,20 @@
       sub1.terminate
       sub2.terminate
 
-      thread_count = Thread.list.count
+      MongoPubsub.actor.inspect # ping to allow subscriptions to terminate
+      GC.start
+
+      thread_list = Thread.list
 
       sub1 = described_class.subscribe('channel1') {|msg| }
       sub2 = described_class.subscribe('channel1') {|msg| }
       sub1.terminate
       sub2.terminate
 
+      MongoPubsub.actor.inspect # ping to allow subscriptions to terminate
       GC.start
-      expect(thread_count == Thread.list.count).to be_truthy
+
+      expect(Thread.list).to eq thread_list
```

Comment:
This started to fail because of `dead` Celluloid threads for some reason now:

```
  1) MongoPubsub.publish cleanups threads
     Failure/Error: expect(Thread.list).to eq thread_list
       
       
       Diff:
       @@ -16,5 +16,6 @@
         #<Celluloid::Thread:0x007fc94c04d518@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x00000003eff080@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x007fc94c04ec88@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       - #<Celluloid::Thread:0x007fc94c057158@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>]
       + #<Celluloid::Thread:0x007fc94c057158@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       + #<Celluloid::Thread:0x000000018a2b08@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 dead>]
       
     # ./spec/services/mongo_pubsub_spec.rb:112:in `block (3 levels) in <top (required)>'
```

The spec is now "leaking" a dead thread for some reason.

Reply:
Well, it also seems to go both ways, time time it failed because some `dead` threads went away:

```
       Diff:
       @@ -16,7 +16,5 @@
         #<Celluloid::Thread:0x007f3b80199ea8@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x007f3b8018b6f0@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x007f3b80188428@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       - #<Celluloid::Thread:0x007f3b803cb528@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       - #<Celluloid::Thread:0x000000013e0840@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 dead>,
       - #<Celluloid::Thread:0x000000013fe818@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 dead>]
       + #<Celluloid::Thread:0x007f3b803cb528@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>]
```

There must be something very subtle going on, possibly related to how the `described_class.subscribe('...')` in the specs no longer go through the `@supervisor.actors` call, and the new `Celluloid::Actors[...]` lookup is now faster... fix is probably to just ignore the `dead` threads.

Reply:
Nope... still flaky, there's `dead` threads even with `select{|t| t.alive?}`, must be a race...

Without the `MongoPubsub.actor.inspect` this fails more or less every time now, and the threads are still `run` at that point, not `dead`:

```
       Diff:
       @@ -16,5 +16,7 @@
         #<Celluloid::Thread:0x000000041a0230@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x007f56c404c488@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
         #<Celluloid::Thread:0x007f56c404eaa8@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       - #<Celluloid::Thread:0x0000000655cd90@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>]
       + #<Celluloid::Thread:0x0000000655cd90@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 sleep>,
       + #<Celluloid::Thread:0x00000001cd5e78@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 run>,
       + #<Celluloid::Thread:0x00000001e02350@/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/group/spawner.rb:47 run>]
```